builtin/k8s/platform.go

```diff
@@ -296,318 +294,392 @@ func (p *Platform) resourceDeploymentStatus(
 	return nil
 }
 
-// resourceDeploymentCreate creates the Kubernetes deployment.
-func (p *Platform) resourceDeploymentCreate(
-	ctx context.Context,
+func configureK8sContainer(
+	name string,
+	image string,
+	ports []*Port,
+	envVars map[string]string,
+	probe *Probe,
+	probePath string,
+	cpu *ResourceConfig,
+	memory *ResourceConfig,
+	resources map[string]string,
+	command *[]string,
+	args *[]string,
+	scratchSpace []string,
+	volumes []corev1.Volume,
 	log hclog.Logger,
-	src *component.Source,
-	img *docker.Image,
-	deployConfig *component.DeploymentConfig,
-	ui terminal.UI,
-
-	result *Deployment,
-	state *Resource_Deployment,
-	csinfo *clientsetInfo,
-	sg terminal.StepGroup,
-) error {
-	// Prepare our namespace and override if set.
-	ns := csinfo.Namespace
-	if p.config.Namespace != "" {
-		ns = p.config.Namespace
+) (*corev1.Container, error) {
+	// If the user is using the latest tag, then don't specify an overriding pull policy.
+	// This by default means kubernetes will always pull so that latest is useful.
+	splitImage := strings.Split(image, ":")
+	var pullPolicy corev1.PullPolicy
+	if len(splitImage) == 1 {
+		// If no tag is set, docker will default to "latest", and we always want to pull.
+		pullPolicy = corev1.PullAlways
+	} else if len(splitImage) == 2 {
+		tag := splitImage[1]
+		if tag == "latest" {
+			// Always pull a latest image, so the k8s workers don't use their local cache and ignore new changes.
+			pullPolicy = corev1.PullAlways
+		} else {
+			// A tag is present, we can use k8s worker caching
+			pullPolicy = corev1.PullIfNotPresent
+		}
+	} else {
```

Comment:
I'm not sure we should fail here, we can probably let K8S do its thing if the image is wrong and let it surface the error instead. The alternative is getting a really good regex to match container registry URLs in the image name which might be a pain. This is one reason why the integration tests are failing. An image name can have a container registry in front of it:

```shell
# a local registry
localhost:5000/sinatra:latest
# or a live one
docker.io/webapp:1.2.3
```